### PR TITLE
Fix #470 non adapted sample indices in gradients during sorting

### LINKED ISSUE #470: `sort` does not adapt samples indices in gradients to the sorted samples in the block the gradients are attached to
```python
import numpy as np
import metatensor
from metatensor import TensorMap, TensorBlock, Labels
block_1 = TensorBlock(
    values=np.array([[3, 5], [1, 2]]),
    samples=Labels(["s"], np.array([[2], [0]])),
    components=[],
    properties=Labels(["p"], np.array([[0], [1]])),
)
block_1.add_gradient(
    parameter="g",
    gradient=TensorBlock(
        values=np.array(
            [[[11, 10],
              [13, 12]],
             [[15, 14],
              [11, 10]],
             [[13, 12],
              [15, 14]]]
        ),
        samples=Labels(
            ["sample", "g"],
            np.array([[0, 1], [0, 2], [1, 1]]),
        ),
        components=[
            Labels(["c"], np.array([[0], [1]])),
        ],
        properties=Labels(["p"], np.array([[0], [1]])),
    ),
)
keys = Labels(names=["_"], values=np.array([[0]]))
tensor = TensorMap(keys, [block_1])
sorted_tensor = metatensor.sort(tensor)
print("tensor")
print(tensor[0].samples)
print(tensor[0].gradient('g').samples)
print()
print("sorted tensor")
print(sorted_tensor[0].samples)
print(sorted_tensor[0].gradient('g').samples)
```
Output:
```bash
tensor
Labels(
    s
    2
    0
)
Labels(
    sample  g
      0     1
      0     2
      1     1
)

sorted tensor
Labels(
    s
    0
    2
)
Labels(
    sample  g
      0     1
      0     2
      1     1
)
```
The gradients assignment to the samples gets swapped in this example. I think this would have been caught by finite difference tests https://github.com/lab-cosmo/metatensor/issues/84. So I look how feasible this is to do for sort.

EDIT: Realised that the sort operation is not really suitable for gradcheck. Will just add regular test

### PR DESCRIPTION
**Pull request by agoscinski**
Fixes #470

- Splitting sorted tensors into an ascending and descending so the fix can be more robustly tested
- Adding dispatched arange_like to create inverse mapping for reasigning samples in gradients

Waiting till https://github.com/lab-cosmo/metatensor/pull/479 is merged to also add finite difference test

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [x] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
📚 Documentation preview 📚: https://metatensor--480.org.readthedocs.build/en/480/

### CODE REVIEW

File: python/metatensor-operations/metatensor/operations/_dispatch.py [hunk 1]
```diff
@@ -715,3 +715,21 @@ def zeros_like(array, shape: Optional[List[int]] = None, requires_grad: bool = F
         return np.zeros_like(array, shape=shape, subok=False)
     else:
         raise TypeError(UNKNOWN_ARRAY_TYPE)
+
+
+def arange_like(array, start, end, requires_grad: bool = False):
```
Comment (Luthaf) [inline python/metatensor-operations/metatensor/operations/_dispatch.py:720]: I don't see the point of requires_grad here, since this is intended to work with integer values, right?

Comment (agoscinski) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:720]: yes makes sense, I just copied pasted from zeros_like

Comment (agoscinski) [reply python/metatensor-operations/metatensor/operations/_dispatch.py]: removed arange completely and used int_like as you suggested in another PR

File: python/metatensor-operations/metatensor/operations/sort.py [hunk 2]
```diff
@@ -4,6 +4,79 @@
...
 
 
+def _sort_single_gradient_block(
+    block: TensorBlock,
+    gradient_block: TensorBlock,
+    axes: List[str],
+    descending: bool,
+) -> TensorBlock:
+    """
+    Sorts a single TensorBlock without the user input checking and sorting of gradients
```
Comment (Luthaf) [inline python/metatensor-operations/metatensor/operations/sort.py:14]:
```suggestion
    Sorts a single gradient TensorBlock. This is different from `_sort_single_block` because we need to update the sample differently (since gradient samples are pointers into the values samples)
```

File: python/metatensor-operations/metatensor/operations/sort.py [hunk 3]
```diff
@@ -4,6 +4,79 @@
...
+    components_values = []
+    for component in gradient_block.components:
+        component_names.append(component.names)
+        components_values.append(component.values)
+
+    property_names = gradient_block.properties.names
+    properties_values = gradient_block.properties.values
+
+    values = gradient_block.values
+    if "samples" in axes:
```
Comment (Luthaf) [inline python/metatensor-operations/metatensor/operations/sort.py:30]: just to be clear, you don't re-sort the "sample" column but update it according to the values samples sorting. Does this re-sort the other columns in the samples?

File: python/metatensor-operations/tests/sort.py [hunk 4]
```diff
@@ -63,11 +62,12 @@ def tensor():
     )
     # fmt: on
     keys = Labels(names=["key_1", "key_2"], values=np.array([[1, 0], [0, 0]]))
+    # block order is descending
     return TensorMap(keys, [block_2, block_1])
 
 
 @pytest.fixture
-def tensor_sorted():
+def tensor_sorted_ascending():
```
Comment (Luthaf) [inline python/metatensor-operations/tests/sort.py:70]: maybe a small docstring explaining the difference with the other fixture?

File: python/metatensor-operations/tests/sort.py [hunk 5]
```diff
@@ -124,31 +126,87 @@ def tensor_sorted():
     return TensorMap(keys, [block_1, block_2])
 
 
-def test_sort(tensor, tensor_sorted):
+@pytest.fixture
+def tensor_sorted_descending():
```
Comment (Luthaf) [inline python/metatensor-operations/tests/sort.py:130]: same!

File: python/metatensor-operations/tests/sort.py [hunk 6]
```diff
@@ -124,31 +126,87 @@ def tensor_sorted():
...
+    block_1 = TensorBlock(
+        values=np.array([[3, 5], [1, 2]]),
+        samples=Labels(["s"], np.array([[2], [0]])),
+        components=[],
+        properties=Labels(["p"], np.array([[1], [0]])),
+    )
+
+    # fmt: off
```
Comment (Luthaf) [inline python/metatensor-operations/tests/sort.py:138]: this should not be needed

File: python/metatensor-torch/tests/operations/sort.py [hunk 7]
```diff
@@ -65,11 +65,12 @@ def tensor():
     )
     # fmt: on
     keys = Labels(names=["key_1", "key_2"], values=torch.tensor([[1, 0], [0, 0]]))
+    # block order is descending
     return TensorMap(keys, [block_2, block_1])
 
 
 @pytest.fixture
-def tensor_sorted():
+def tensor_sorted_ascending():
```
Comment (Luthaf) [inline python/metatensor-torch/tests/operations/sort.py:73]: same about the docstring

File: python/metatensor-operations/tests/sort.py (hunk 6 above)
Comment (agoscinski) [reply python/metatensor-operations/tests/sort.py:138]: not needed but black makes this really hard to read so I turned it off

File: python/metatensor-operations/metatensor/operations/sort.py (hunk 3 above)
Comment (agoscinski) [reply python/metatensor-operations/metatensor/operations/sort.py:30]: I updated the in-code doc a little bit

File: python/metatensor-operations/metatensor/operations/sort.py (hunk 2 above)
Comment (agoscinski) [reply python/metatensor-operations/metatensor/operations/sort.py]: updated this doc and embedded suggested change into it

File: python/metatensor-operations/metatensor/operations/sort.py (hunk 3 above)
Comment (Luthaf) [reply python/metatensor-operations/metatensor/operations/sort.py:33]: sounds good! I had a hard time following some of this

File: python/metatensor-operations/tests/sort.py (hunk 6 above)
Comment (agoscinski) [reply python/metatensor-operations/tests/sort.py:146]: This have strong feelings on. This sort test is already hard to understand as it is, and the black formatting made it much harder.

Comment (agoscinski) [reply python/metatensor-operations/tests/sort.py:146]: Unlike other tests, the order must be much easier to read so one can follow what is how moved. I added comments why it is turned off.
